src/ore/src/task.rs

```diff
@@ -74,72 +76,114 @@ impl<T> Future for AbortOnDropHandle<T> {
+/// - `JoinHandleExt::abort_and_wait`: abort the task and wait for it to be finished.
+/// - `into_tokio_handle`: turn it into an ordinary tokio `JoinHandle`.
+#[derive(Debug)]
+pub struct JoinHandle<T>(TokioJoinHandle<T>);
 
-    /// Awaits the `JoinHandle`, resuming the unwind if the task panicked,
-    /// and panicking if the task was cancelled.
-    ///
-    /// If abort-on-drop is on, then this method will only
-    /// panic on cancelled tasks.
+impl<T> Future for JoinHandle<T> {
+    type Output = Result<T, JoinError>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        self.0.poll_unpin(cx)
+    }
+}
+
+impl<T> JoinHandle<T> {
+    /// Create an [`AbortOnDropHandle`] from this [`JoinHandle`].
+    pub fn abort_on_drop(self) -> AbortOnDropHandle<T> {
+        AbortOnDropHandle(self.0)
+    }
+
+    /// Checks if the task associated with this [`JoinHandle`] has finished.a
+    pub fn is_finished(&self) -> bool {
+        self.0.is_finished()
+    }
+
+    /// Checks if the task associated with this [`JoinHandle`] has finished.a
+    pub fn into_tokio_handle(self) -> TokioJoinHandle<T> {
+        self.0
+    }
+
+    // Note: adding an `abort(&self)` method here is incorrect, please see `unpack_join_result`.
+}
+
+/// Extension methods for [`JoinHandle`] and [`AbortOnDropHandle`].
+#[async_trait::async_trait]
+pub trait JoinHandleExt<T>: Future<Output = Result<T, JoinError>> {
+    /// Waits for the task to finish, resuming the unwind if the task panicked.
     ///
-    /// This method will also yield to the runtime
+    /// Because this takes ownership of `self`, and [`JoinHandle`] and
+    /// [`AbortOnDropHandle`] don't offer `abort` methods, this can avoid
+    /// worrying about aborted tasks.
     async fn wait_and_assert_finished(self) -> T;
+
+    /// Abort the `Jo
+    async fn abort_and_wait(self);
 }
 
 async fn unpack_join_result<T>(res: Result<T, JoinError>) -> T {
     match res {
         Ok(val) => val,
         Err(err) => match err.try_into_panic() {
             Ok(panic) => std::panic::resume_unwind(panic),
-            Err(e) => {
-                // Yield explicitly to the tokio runtime. The runtime
-                // can only shutdown each core thread when the running
-                // task yields. This means we could panic here if the `JoinHandle`
-                // is cancelled because it is owned by a separate core thread,
-                // which is already shutdown. We don't want to panic
-                // spuriously when we are already shutting down.
-                tokio::task::yield_now().await;
-                panic!("task expected to complete was cancelled: {}", e);
+            Err(_e) => {
+                // Because `JoinHandle` and `AbortOnDropHandle` don't
+                // offer `abort` method, this can only happen if the runtime is
+                // shutting down, which means this `pending` won't cause a deadlock.
```

Comment:
```suggestion
                // shutting down, which means this `pending` won't cause a deadlock
                // because Tokio drops all outstanding futures on shutdown.
                // (In multi-threaded runtimes, not all threads drop futures simultaneously,
                // so it is possible for a future on one thread to observe the drop of a future
                // on another thread, before it itself is dropped.)
```

Like one more phrase of explanation would be great. Is what I suggested correct?

Reply:
yes!